app/src/main/java/com/amaze/filemanager/filesystem/Operations.java

```diff
@@ -441,6 +441,63 @@ public static void rename(
 
       private final DataUtils dataUtils = DataUtils.getInstance();
 
+      private final Boolean isCaseSensitiveRename =
+              !oldFile.getSimpleName().equals(newFile.getSimpleName()) &&
+              oldFile.getSimpleName().equalsIgnoreCase(newFile.getSimpleName());
+
+      // random string that is appended to file to prevent name collision
+      private static final String TEMP_FILE_EXT = "qwerty";
+
+      private boolean defaultRename(
+          @NonNull HybridFile oldFile,
+          @NonNull HybridFile newFile
+      ) {
+        File file = new File(oldFile.getPath());
+        File file1 = new File(newFile.getPath());
+        boolean result = false;
+
+        switch (oldFile.getMode()) {
+          case FILE:
+            int mode = checkFolder(file.getParentFile(), context);
+            if (mode == 2) {
+              errorCallBack.launchSAF(oldFile, newFile);
+            } else if (mode == 1 || mode == 0) {
+              try {
+                RenameOperation.renameFolder(file, file1, context);
+                result = true;
+              } catch (ShellNotRunningException e) {
+                LOG.warn("failed to rename file in local filesystem", e);
+              }
+              boolean a = !file.exists() && file1.exists();
+              if (!a && rootMode) {
+                try {
+                  RenameFileCommand.INSTANCE.renameFile(file.getPath(), file1.getPath());
+                  result = true;
+                } catch (ShellNotRunningException e) {
+                  LOG.warn("failed to rename file in local filesystem", e);
+                }
+                oldFile.setMode(OpenMode.ROOT);
+                newFile.setMode(OpenMode.ROOT);
+                a = !file.exists() && file1.exists();
+              }
+              errorCallBack.done(newFile, a);
```

Comment:
Need a fallback mechanism in case the second rename fails, fallback should attempt to rename back to original file name.

Reply:
I added a simple check that attempt to rename back to original file name, not sure if it is sufficient.

Reply:
Thanks, changes seem fine. It would be a good practice to LOG.warn the inner else condition in case second rename fails.
Also, codacy does seem to complain can you fix the issue:

Check warning on line 673 in app/src/main/java/com/amaze/filemanager/filesystem/Operations.java

Codacy Production
/ Codacy Static Code Analysis
app/src/main/java/com/amaze/filemanager/filesystem/Operations.java#L673
These nested if statements could be combined

Reply:
I added a log message and combined those if statements.